extendr-macros/src/wrappers.rs

```diff
@@ -120,6 +145,43 @@ pub fn make_function_wrappers(
             });)
         })
         .unwrap_or_default();
+
+    // figure out if &Self / &mut Self / inside of impl block &ImplType / &mut ImplType is used!
+    let return_is_ref_self = {
+        match sig.output {
+            syn::ReturnType::Default => false,
+            syn::ReturnType::Type(_, ref return_type) => match return_type.as_ref() {
+                // matches `-> Self`
+                // Type::Path(type_path) => type_path.path.is_ident("Self"),
+                // matches `-> &Self` / `-> &mut Self`
+                Type::Reference(ref reference_type) => {
+                    if let Type::Path(path) = reference_type.elem.as_ref() {
+                        path.path.is_ident("Self")
```

Comment:
Is it possible to reorder the comments? I'm not sure which is applying to what. I _think_ I've got it right? Except for the `Type::Path(type_path) => type_path.path.is_ident("Self"),` seems like it might be commented out as old code? 

```suggestion
             // matches `-> Self`
            syn::ReturnType::Type(_, ref return_type) => match return_type.as_ref() {
                Type::Reference(ref reference_type) => {
                    // Type::Path(type_path) => type_path.path.is_ident("Self"),
                    // matches `-> &Self` / `-> &mut Self`
                    if let Type::Path(path) = reference_type.elem.as_ref() {
                        path.path.is_ident("Self")
```

Reply:
Not really old code, but it is here I signal that taken `self: Self` is not handled, and will be `false`, and it is not allowed.
Also, now that I think about it, people might write

```rs

struct Model;

#[extendr]
impl Model {
    fn take(self: Model) {
        
    }
}
```
... which is frightening to think about. I should find a solution for this...